src/goLanguageServer.ts

```diff
@@ -6,92 +6,119 @@
 import { promptForMissingTool, promptForUpdatingTool } from './goInstallTools';
-import { parseLiveFile } from './goLiveErrors';
-import { GO_MODE } from './goMode';
-import { GoDocumentSymbolProvider } from './goOutline';
 import { getToolFromToolPath } from './goPath';
-import { GoReferenceProvider } from './goReferences';
-import { GoRenameProvider } from './goRename';
-import { GoSignatureHelpProvider } from './goSignature';
-import { GoCompletionItemProvider } from './goSuggest';
-import { GoWorkspaceSymbolProvider } from './goSymbol';
 import { getTool, Tool } from './goTools';
-import { GoTypeDefinitionProvider } from './goTypeDefinition';
 import { getBinPath, getCurrentGoPath, getGoConfig, getToolsEnvVars } from './util';
 
 interface LanguageServerConfig {
+	serverName: string;
+	path: string;
 	enabled: boolean;
 	flags: string[];
+	env: any;
 	features: {
 		diagnostics: boolean;
 		documentLink: boolean;
 	};
 	checkForUpdates: boolean;
 }
 
-// registerLanguageFeatures registers providers for all the language features.
-// It looks to either the language server or the standard providers for these features.
-export async function registerLanguageFeatures(ctx: vscode.ExtensionContext) {
+// Global variables used for management of the language client.
+// They are global so that the server can be easily restarted with
+// new configurations.
+let languageClient: LanguageClient;
+let languageServerDisposable: vscode.Disposable;
+let latestConfig: LanguageServerConfig;
+let serverOutputChannel: vscode.OutputChannel;
+
+// startLanguageServer starts the language server (if enabled), returning
+// true on success.
+export async function registerLanguageFeatures(ctx: vscode.ExtensionContext): Promise<boolean> {
 	// Subscribe to notifications for changes to the configuration of the language server.
 	ctx.subscriptions.push(vscode.workspace.onDidChangeConfiguration((e) => watchLanguageServerConfiguration(e)));
 
 	const config = parseLanguageServerConfig();
-
-	// If the user has not enabled the language server,
-	// register the default language features and return.
 	if (!config.enabled) {
-		registerUsualProviders(ctx);
-		return;
+		return false;
 	}
 
-	// The user has opted into the language server.
-	const languageServerToolPath = getLanguageServerToolPath();
-	const toolName = getToolFromToolPath(languageServerToolPath);
-	if (!toolName) {
-		// language server binary is not installed yet.
-		// Return immediately. The information messages such as
-		// offering to install missing tools, and suggesting to
-		// reload the window after installing the language server
-		// should be presented by now.
-		return;
+	// Support a command to restart the language server, if it's enabled.
+	ctx.subscriptions.push(vscode.commands.registerCommand('go.languageserver.restart', () => {
+		return startLanguageServer(ctx, parseLanguageServerConfig());
+	}));
+
+	const tool = getTool(config.serverName);
+	if (!tool) {
+		return false;
```

Comment:
`getTool()` works for the predefined set of tools we work with.
`config.servername` can potentially have a value that is not in the predefined set of tools if the user has used `go.alternateTools` to point `gopls` to something else
In such case, we will end up returning false here

Reply:
I would revert the check here to
```js
if (config.serverName === 'gopls') {
    // call shouldUpdateLanguageServer and prompt to update if needed
}
```

Reply:
Done.